pkg/lib/aws/servicequotas.go

```diff
@@ -103,3 +110,156 @@ func (c *Client) VerifyInstanceQuota(instanceType string, requiredOnDemandInstan
 
 	return nil
 }
+
+func (c *Client) VerifyNetworkQuotas(requiredInternetGateways int, requiredNATGatewaysPerAZ int, highlyAvailableNATGateway bool, requiredVPCs int, availabilityZones strset.Set) error {
+	ec2QuotaCodeToValueMap := map[string]int{
+		_elasticIPsQuotaCode: 0, // elastic IP quota code
+	}
+	vpcQuotaCodeToValueMap := map[string]int{
+		_internetGatewayQuotaCode: 0, // internet gw quota code
+		_natGatewayQuotaCode:      0, // nat gw quota code
+		_vpcQuotaCode:             0, // vpc quota code
+	}
+
+	err := c.ServiceQuotas().ListServiceQuotasPages(
+		&servicequotas.ListServiceQuotasInput{
+			ServiceCode: aws.String("ec2"),
+		},
+		func(page *servicequotas.ListServiceQuotasOutput, lastPage bool) bool {
+			if page == nil {
+				return false
+			}
+			for _, quota := range page.Quotas {
+				if quota == nil || quota.QuotaCode == nil || quota.Value == nil {
+					continue
+				}
+				if _, ok := ec2QuotaCodeToValueMap[*quota.QuotaCode]; ok {
+					ec2QuotaCodeToValueMap[*quota.QuotaCode] = int(*quota.Value)
+					return false
+				}
+			}
+			return true
+		},
+	)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	err = c.ServiceQuotas().ListServiceQuotasPages(
+		&servicequotas.ListServiceQuotasInput{
+			ServiceCode: aws.String("vpc"),
+		},
+		func(page *servicequotas.ListServiceQuotasOutput, lastPage bool) bool {
+			if page == nil {
+				return false
+			}
+			for _, quota := range page.Quotas {
+				if quota == nil || quota.QuotaCode == nil || quota.Value == nil {
+					continue
+				}
+				if _, ok := vpcQuotaCodeToValueMap[*quota.QuotaCode]; ok {
+					vpcQuotaCodeToValueMap[*quota.QuotaCode] = int(*quota.Value)
+				}
+			}
+			return true
+		},
+	)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get EIP in use
+	elasticIPsInUse, err := c.ListElasticIPs()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get IGW in use
+	internetGatewaysInUse, err := c.ListInternetGateways()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get NAT GW in use per selected AZ
+	gateways, err := c.DescribeNATGateways()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	subnets, err := c.DescribeSubnets()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	gatewaysSubnetIDs := SubnetIDsFromNATGateways(gateways)
+	azToGatewaysInUse := make(map[string]int)
+	for _, gatewaySubnetID := range gatewaysSubnetIDs {
+		for _, subnet := range subnets {
+			if subnet.SubnetId == nil || subnet.AvailabilityZone == nil {
+				continue
+			}
+			if !availabilityZones.Has(*subnet.AvailabilityZone) {
+				continue
+			}
+			if *subnet.SubnetId == gatewaySubnetID {
+				if _, ok := azToGatewaysInUse[*subnet.AvailabilityZone]; !ok {
+					azToGatewaysInUse[*subnet.AvailabilityZone] = 1
+				} else {
+					azToGatewaysInUse[*subnet.AvailabilityZone]++
+				}
+			}
+		}
+	}
+
+	// get VPC IDs
+	vpcs, err := c.DescribeVpcs()
```

Comment:
nit: this assignment can be moved down to the line above the first place `vpcs` is used